File: pipeline/models/classes.py
"""Pydantic model for classes.json."""

from __future__ import annotations

from pydantic import field_validator

from .common import ClassFeatEntry, StrictModel


class ClassesMetadata(StrictModel):
    """Metadata block at the top of classes.json."""

    description: str
    version: str
    levelsComplete: list[int]
    levelsPending: list[int]


class TrackInfo(StrictModel):
    """A class track showing L1→L5 progression."""

    name: str
    classes: list[str]


class GameClass(StrictModel):
    """A single class entry."""

    id: str
    name: str
    level: int
    track: str | None
    prerequisites: str
    characteristics: list[str]
    skills: list[str]
    feats: list[ClassFeatEntry]
    swordSchools: list[str]
    magicSchools: list[str]
    gunKata: list[str]
    completionBonus: str
    suggestedExits: list[str]

    @field_validator("level")
    @classmethod
    def level_in_range(cls, v: int) -> int:
        if not 1 <= v <= 5:
            msg = f"Level must be 1-5, got {v}"
            raise ValueError(msg)
        return v


class ClassesFile(StrictModel):
    """Top-level structure of classes.json."""

    metadata: ClassesMetadata
    tracks: dict[str, TrackInfo]
    classes: list[GameClass]

File: src/lib/dtd/schemas/classes.ts
/**
 * Zod schemas for class data.
 *
 * Replaces former Pydantic model.
 */

import { z } from "zod";
import { ClassFeatEntry } from "./common.js";

// ---------------------------------------------------------------------------
// Sub-schemas
// ---------------------------------------------------------------------------

export const ClassesMetadata = z
	.object({
		description: z.string(),
		version: z.string(),
		levelsComplete: z.array(z.number().int()),
		levelsPending: z.array(z.number().int()),
	})
	.strict();
export type ClassesMetadata = z.infer<typeof ClassesMetadata>;

export const TrackInfo = z
	.object({
		name: z.string(),
		classes: z.array(z.string()),
	})
	.strict();
export type TrackInfo = z.infer<typeof TrackInfo>;

export const GameClass = z
	.object({
		id: z.string(),
		name: z.string(),
		level: z.number().int().min(1).max(5),
		track: z.string().nullable(),
		prerequisites: z.string(),
		characteristics: z.array(z.string()),
		skills: z.array(z.string()),
		feats: z.array(ClassFeatEntry),
		swordSchools: z.array(z.string()),
		magicSchools: z.array(z.string()),
		gunKata: z.array(z.string()),
		completionBonus: z.string(),
		suggestedExits: z.array(z.string()),
	})
	.strict();
export type GameClass = z.infer<typeof GameClass>;

// ---------------------------------------------------------------------------
// Top-level schema
// ---------------------------------------------------------------------------

export const ClassesFile = z
	.object({
		metadata: ClassesMetadata,
		tracks: z.record(z.string(), TrackInfo),
		classes: z.array(GameClass),
	})
	.strict();
export type ClassesFile = z.infer<typeof ClassesFile>;
